src/newtondynamics_physics/ospnewton.h

```diff
@@ -0,0 +1,124 @@
+/**
+ * Open Space Program
+ * Copyright © 2019-2021 Open Space Program Project
+ *
+ * MIT License
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+#pragma once
+
+#include <osp/Active/SysPhysics.h>
+#include <osp/Active/ActiveScene.h>
+
+#include <Newton.h>
+
+namespace ospnewton
+{
+
+/**
+ * @brief Associates an entity with a NewtonBody and manages its lifetime
+ */
+struct ACompNwtBody
+{
+
+    ACompNwtBody(NewtonBody const* pBody) noexcept
+     : m_pBody(pBody) { }
+
+    NewtonBody const* body() const noexcept { return m_pBody.get(); }
+
+private:
+    struct Deleter
+    {
+        void operator() (NewtonBody const* pCollision)
+        { NewtonDestroyBody(pCollision); }
+    };
+
+    std::unique_ptr<NewtonBody const, Deleter> m_pBody;
+};
+
+/**
+ * @brief Associates an entity with a NewtonCollision and manages its lifetime
+ */
+struct ACompNwtCollider
+{
+
+    ACompNwtCollider(NewtonCollision const* pCollision) noexcept
+     : m_pCollision(pCollision) { }
+
+    NewtonCollision const* collision() const noexcept
+    { return m_pCollision.get(); }
+
+private:
+    struct Deleter
+    {
+        void operator() (NewtonCollision const* pCollision)
+        { NewtonDestroyCollision(pCollision); }
+    };
+
+    std::unique_ptr<NewtonCollision const, Deleter> m_pCollision;
+};
+
+/**
+ * @brief Represents an instance of a Newton physics world in the scane
+ */
+struct ACtxNwtWorld
+{
+
+    // TODO: not compiling, std::hardware_destructive_interference_size not found?
+    struct alignas(64) PerThread
```

Comment:
You said in discord that this alignas(64) was supposed to represent the size of the CPUs cache line.

Recommend documenting that in the code.

Reply:
What exactly is the behavior of `alignas()`? Do all instances of this struct get a bunch of extra padding and are shifted within their memory footprint to align them? I'm only familiar with struct packing and allocating aligned dynamic memory, I've never heard of this macro before

Reply:
These structs are stored contiguously in `std::vector<PerThread> m_perThread;`, and are accessed among different threads simultaneously. Without proper alignment, they're small enough to fall into the same cache line. This can lead to serious performance degradation from False Sharing, as the cpu cores will end up invalidating each other's caches when they write a new value.

`alignas` specifies alignment requirements, though I'm not entirely familiar with what this really affects. If I'm doing this correctly, then the struct's size will be a multiple of 64, and the standard C++17 allocators will do aligned allocations.

Reply:
Ok, interesting. I wouldn't have expected this to be supported by MSVC since they don't support `std::aligned_alloc` yet but perhaps since they get control over both ends it works out. The build seems to pass, anyway.